File: src/java/org/javaturk/dp/ch05/flyweight/book/Line.java
package org.javaturk.dp.ch05.flyweight.book;

import java.util.ArrayList;
import java.util.List;

public class Line {
	private List<Character> chars;
	private static Character endOfLine = new Character('\n');
	private int emptyPosition;
	private boolean full;
	private int numberOfCharacters = 1;

	public Line(int numberOfCharacters) {
		this.numberOfCharacters = numberOfCharacters;
		chars = new ArrayList<>(numberOfCharacters);
	}

	public boolean add(Character character) {
		if (!full) {
			chars.add(character);
			character.setLine(this);
			character.setPosition(emptyPosition);
			emptyPosition++;
			if(emptyPosition == numberOfCharacters + 1)
				full = true;
			return true;
		} else
			return false;
	}

	public void addEndOfLine() {
		chars.add(endOfLine);
	}

	public List<Character> getChars() {
		return chars;
	}

	public String toString() {
		StringBuilder sb = new StringBuilder();
		for (Character character : chars)
			sb.append(character.getValue());
		return sb.toString();
	}
}

File: src/typescript/com/gokselkucuksahin/dp/structural-patterns/flyweight/book/line.ts
import {Character} from "./character";

export class Line {
  private chars: Character[];
  private static endOfLine: Character = Character.character_1('\n');
  private emptyPosition: number;
  private full: boolean;
  private numberOfCharacters: number = 1;

  constructor(numberOfCharacters: number) {
    this.numberOfCharacters = numberOfCharacters;
    this.chars = Array<Character>(numberOfCharacters);
  }

  add(character: Character): boolean {
    if (!this.full) {
      this.chars.push(character);
      character.line = this;
      character.position = this.emptyPosition;
      this.emptyPosition++;
      if (this.emptyPosition == this.numberOfCharacters + 1) {
        this.full = true;
      }
      return true;
    } else {
      return false;
    }
  }

  addEndOfLine(): void {
    this.chars.push(Line.endOfLine);
  }

  getChars(): Character[] {
    return this.chars;
  }

  toString(): string {
    return this.getChars().map(x => x.toString()).join('');
  }
}